controllers/druid/drain_historical.go

```diff
@@ -0,0 +1,191 @@
+	appsv1 "k8s.io/api/apps/v1"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+var logger_drain_historical = logf.Log.WithName("drain_historical")
+
+type CoordinatorConfig struct {
+	decommissioningNodes []string
+}
+
+func checkDrainStatus(d DruidClient, start int32, end int32, podNames []string) (bool, error) {
+	historicalUsageStats, err := d.GetHistoricalUsage()
+	if err != nil {
+		return false, err
+	}
+	for it := start; it <= end; it++ {
+		if historicalUsageStats[podNames[it]] != 0 {
+			return false, nil
+		}
+	}
+	return true, nil
+}
+
+func patchUpdateStrategy(sdk client.Client, m *v1alpha1.Druid, nodeSpec *v1alpha1.DruidNodeSpec, strategy string, emitEvent EventEmitter) error {
+	updatedObj := nodeSpec.DeepCopy()
+
+	newStrategy := appsv1.StatefulSetUpdateStrategy{
+		Type: appsv1.OnDeleteStatefulSetStrategyType,
+	}
+
+	updatedObj.UpdateStrategy = &newStrategy
+
+	if err := writers.Patch(context.TODO(), sdk, m, updatedObj, false, client.MergeFrom(nodeSpec), emitEvent); err != nil {
+		return err
+	} else {
+		msg := fmt.Sprintf("Successfully Patched %s with %s", nodeSpec.UpdateStrategy, onDelete)
+		logger_drain_historical.Info(msg, "name", m.Name, "namespace", m.Namespace)
+		return nil
+	}
+}
+
+func scaleStatefulSet(sdk client.Client, m *v1alpha1.Druid, nodeSpec *v1alpha1.DruidNodeSpec, strategy string, emitEvent EventEmitter, scaleUpCount int32) error {
+	updatedObj := nodeSpec.DeepCopy()
+	newReplica := int32(m.Status.Historical.Replica + scaleUpCount)
+	updatedObj.Replicas = newReplica
+	if err := writers.Patch(context.TODO(), sdk, m, updatedObj, false, client.MergeFrom(nodeSpec), emitEvent); err != nil {
+		return err
+	} else {
+		msg := fmt.Sprintf("Successfully Patched replicas with %d", newReplica)
+		logger_drain_historical.Info(msg, "name", m.Name, "namespace", m.Namespace)
+		return nil
+	}
+}
+
+func deployHistorical(sdk client.Client, m *v1alpha1.Druid, nodeSpec *v1alpha1.DruidNodeSpec, nodeSpecUniqueStr string, emitEvent EventEmitter, batchSize int32, baseUrl string) error {
+	// patch the updateStrategy with onDelete
+	err := patchUpdateStrategy(sdk, m, nodeSpec, onDelete, emitEvent)
+	if err != nil {
+		return err
+	}
+	// sts first scale up the pods by a batchsize and then drain the older pods
+	if m.Status.Historical.CurrentBatch == 0 {
+		m.Status.Historical = v1alpha1.HistoricalStatus{}
+		m.Status.Historical.Replica = nodeSpec.Replicas
+		m.Status.Historical.DecommissionedPods = []string{}
+
+		err := scaleStatefulSet(sdk, m, nodeSpec, onDelete, emitEvent, batchSize)
+		if err != nil {
+			return err
+		}
+
+		obj, err := readers.Get(context.TODO(), sdk, nodeSpecUniqueStr, m, func() object { return &appsv1.StatefulSet{} }, emitEvent)
+		if err != nil {
+			return err
+		}
+
+		if obj.(*appsv1.StatefulSet).Status.ReadyReplicas != obj.(*appsv1.StatefulSet).Status.CurrentReplicas {
```

Comment:
Also add status.updateRevision and state.currentRevision check to ensure that there is a deployment and not just unhealthy pods.